### PR TITLE
Task cgroup limit models and config

### PR DESCRIPTION
**Pull request by petderek**
### Summary

Adds model and config for task limits. 

### Implementation details
* Limit values set at task level
* Capabilities code enhanced to check for docker api 1.22
* Config enhanced to allow users to disable features

### Testing
- [x] Builds on Linux (`make release`)
- [x] Builds on Windows (`go build -out amazon-ecs-agent.exe ./agent`)
- [x] Unit tests on Linux (`make test`) pass
- [x] Unit tests on Windows (`go test -timeout=25s ./agent/...`) pass
- [ ] Integration tests on Linux (`make run-integ-tests`) pass
- [x] Integration tests on Windows (`.\scripts\run-integ-tests.ps1`) pass
- [x] Functional tests on Linux (`make run-functional-tests`) pass
- [x] Functional tests on Windows (`.\scripts\run-functional-tests.ps1`) pass

New tests cover the changes: yes

### Licensing

This contribution is under the terms of the Apache 2.0 License: yes

### CODE REVIEW

File: agent/engine/docker_task_engine.go [hunk 1]
```diff
@@ -830,6 +832,15 @@ func (engine *DockerTaskEngine) Capabilities() []string {
 		}
 	}
 
+	if engine.cfg.TaskResourceLimitEnabled {
+		if _, ok := supportedVersions[dockerclient.Version_1_22]; ok {
+			capabilities = append(capabilities, capabilityPrefix+capabilityTaskResourceLimit)
```
Comment (vsiddharth) [inline agent/engine/docker_task_engine.go]: We need to ensure that we register this capability only on linux.

File: agent/api/task.go [hunk 2]
```diff
@@ -65,7 +65,11 @@ type Task struct {
 	Containers []*Container
 	// Volumes are the volumes for the task
 	Volumes []TaskVolume `json:"volumes"`
-
+	// CPULimit is a task-level limit for compute resources. A value of 1 means that
+	// the task may access 100% of 1 vCPU on the instance
+	CPULimit float64 `json:"CPULimit"`
```
Comment (vsiddharth) [inline agent/api/task.go]: Should we have `omitempty` as these aren't mandatory

**Comment by vsiddharth** [comment]
@petderek Could you please update the summary block's testing section ?

File: agent/engine/docker_task_engine.go (hunk 1 above)
Comment (petderek) [inline agent/engine/docker_task_engine.go]: TaskResourceLimitEnabled will always be overridden to false on windows.

File: agent/api/task.go (hunk 2 above)
Comment (petderek) [inline agent/api/task.go]: Yep! I'll fix that.

**Comment by petderek** [comment]
> ...testing...

I need to verify the integ tests on linux. I'll keep the summary block updated.

> What happens when old agent reads new state file ?

I haven't addressed that case yet. If `omitempty` is set it should be fine because I didn't change any of the required structures, but I can look at that. I'm also concerned about agent creating a hierarchy after it reads an old state file -- but I'm not addressing that second situation in this PR.

File: agent/api/task.go [hunk 3]
```diff
@@ -65,7 +65,11 @@ type Task struct {
 	Containers []*Container
 	// Volumes are the volumes for the task
 	Volumes []TaskVolume `json:"volumes"`
-
+	// CPULimit is a task-level limit for compute resources. A value of 1 means that
```
Comment (samuelkarp) [inline agent/api/task.go]: Should this be `VCPULimit`?

File: agent/config/config_test.go [hunk 4]
```diff
@@ -124,41 +110,33 @@ func TestEnvironmentConfig(t *testing.T) {
 	assert.Equal(t, (90 * time.Second), conf.TaskCleanupWaitDuration)
 }
 
-func TestTrimWhitespace(t *testing.T) {
+func TestTrimWhitespaceWhenCreating(t *testing.T) {
+	os.Setenv("AWS_DEFAULT_REGION", "us-west-2")
```
Comment (samuelkarp) [inline agent/config/config_test.go]: Why was this added?

File: agent/config/config_windows.go [hunk 5]
```diff
@@ -84,4 +84,7 @@ func (config *Config) platformOverrides() {
 		}
 		config.ReservedPorts = append(config.ReservedPorts, httpPort)
 	}
+
+	// ensure TaskResourceLimit is disabled
+	config.TaskResourceLimitEnabled = false
```
Comment (samuelkarp) [inline agent/config/config_windows.go]: Can you add this to `DefaultConfig()` above too?

File: agent/engine/docker_task_engine.go [hunk 6]
```diff
@@ -830,6 +832,15 @@ func (engine *DockerTaskEngine) Capabilities() []string {
 		}
 	}
 
+	if engine.cfg.TaskResourceLimitEnabled {
```
Comment (samuelkarp) [inline agent/engine/docker_task_engine.go]: Can you add a TODO here talking about ensuring compatibility with existing, already-running tasks (upgrade scenario)?  If there are already tasks running without a cgroup parent, we need to disable this feature.

Comment (petderek) [reply agent/engine/docker_task_engine.go]: I'm tracking that separately; I will also add a TODO

File: agent/api/task.go (hunk 3 above)
Comment (petderek) [reply agent/api/task.go]: I don't know. From agent's perspective its just applying a period and a quota. I'll change it if you think its more clear.

File: agent/config/config_test.go (hunk 4 above)
Comment (petderek) [reply agent/config/config_test.go]: Which one?

* `TestTrimWhitespace` was testing two different things, so I split it up.
* `os.Setenv` is there because the test was getting mad without the region explicitly set

File: agent/config/types.go [hunk 7]
```diff
@@ -109,6 +109,9 @@ type Config struct {
 	// tasks with IAM Roles.
 	TaskIAMRoleEnabled bool
 
+	// TaskResourceLimitEnabled specifies if Agent can launch a task with a hierarchical cgroup
+	TaskResourceLimitEnabled bool
```
Comment (aaithal) [inline agent/config/types.go]: This requires a change to `config.String()` as well.

File: agent/engine/docker_task_engine.go [hunk 8]
```diff
@@ -45,6 +45,7 @@ const (
 	capabilityPrefix             = "com.amazonaws.ecs.capability."
 	capabilityTaskIAMRole        = "task-iam-role"
 	capabilityTaskIAMRoleNetHost = "task-iam-role-network-host"
+	capabilityTaskResourceLimit  = "task-resource-limit"
```
Comment (aaithal) [inline agent/engine/docker_task_engine.go]: Can you modify this to be `task-cpu-mem-limit`? `resource` has a pretty broad definition and seems like we're overloading it here. This applies to other names as well, let's make it very explicit by modifying `TaskResourceLimit` to be `TaskCPUMemLimit` everywhere.

File: agent/config/config_test.go (hunk 4 above)
Comment (samuelkarp) [reply agent/config/config_test.go]: Why does it require `os.Setenv` now?

File: agent/api/task.go (hunk 3 above)
Comment (samuelkarp) [reply agent/api/task.go]: I'd like it to match whatever we're going to have in the task definition.  If we don't know that yet, go ahead with either one and then we can fix it once we have an answer.

File: agent/config/config_test.go (hunk 4 above)
Comment (petderek) [reply agent/config/config_test.go]: I looked. The tests are depending on envs being set in other tests. I don't like that. I'll fix it.

File: agent/engine/docker_task_engine.go (hunk 8 above)
Comment (vsiddharth) [reply agent/engine/docker_task_engine.go]: Given that memory is optional, should we split this up into two distinct capabilities ?

Comment (aaithal) [reply agent/engine/docker_task_engine.go]: @vsiddharth the capabilities do not reflect fields that are set in tasks. They are a way of letting agent register if it's capable of handling "certain things". In this particular scenario, it can handle both "cpu" and "memory" at the "task" level. Since we'll be releasing support for both of these with the same agent version, we don't need them to be 2 distinct capabilities.

Comment (vsiddharth) [reply agent/engine/docker_task_engine.go]: @aaithal thanks for the explanation.

File: agent/config/config_test.go [hunk 9]
```diff
@@ -211,170 +188,146 @@ func TestInvalideValueDockerStopTimeout(t *testing.T) {
...
-	}
+	assert.NoError(t, err)
+	assert.Equal(t, cfg.NumImagesToDeletePerCycle, DefaultNumImagesToDeletePerCycle, "Wrong value for NumImagesToDeletePerCycle")
+}
 
-	if cfg.NumImagesToDeletePerCycle != DefaultNumImagesToDeletePerCycle {
-		t.Errorf("Wrong value for NumImagesToDeletePerCycle: %v", cfg.NumImagesToDeletePerCycle)
-	}
+func TestTaskResourceLimitsOverride(t *testing.T) {
+	defer os.Clearenv()
```
Comment (aaithal) [inline agent/config/config_test.go]: I'm not a fan of all the `os.Clearenv` directives here. Can we reduce the scope of the cleanup to be just 
```
defer os.Unsetenv("SOME_SPECIFIC_THING_1")
defer os.Unsetenv("SOME_SPECIFIC_THING_2_IF_NEEDED")
``` 
? My main concern is that it might have some unintended consequences for other unit tests.

File: agent/engine/docker_task_engine.go [hunk 10]
```diff
@@ -830,6 +832,16 @@ func (engine *DockerTaskEngine) Capabilities() []string {
 		}
 	}
 
+	// TODO we also need to disable this if agent already managing older tasks
+	if engine.cfg.TaskCPUMemLimit {
+		if _, ok := supportedVersions[dockerclient.Version_1_22]; ok {
+			capabilities = append(capabilities, capabilityPrefix+capabilityTaskCPUMemLimit)
+		} else {
+			seelog.Warn("Task CPU + Mem Limit disabled due to unsupported Docker version")
```
Comment (aaithal) [inline agent/engine/docker_task_engine.go]: can this be `"Setting task CPU and Memory Limit disabled .."` instead?

File: agent/engine/docker_task_engine.go [hunk 11]
```diff
@@ -771,6 +772,7 @@ func (engine *DockerTaskEngine) State() dockerstate.TaskEngineState {
 //    com.amazonaws.ecs.capability.ecr-auth
 //    com.amazonaws.ecs.capability.task-iam-role
 //    com.amazonaws.ecs.capability.task-iam-role-network-host
+//    com.amazonaws.ecs.capability.task-resource-limit
```
Comment (aaithal) [inline agent/engine/docker_task_engine.go]: nit: this should be `ecs.task-cpu-mem-limit`

File: agent/engine/docker_task_engine.go [hunk 12]
```diff
@@ -830,6 +832,16 @@ func (engine *DockerTaskEngine) Capabilities() []string {
 		}
 	}
 
+	// TODO we also need to disable this if agent already managing older tasks
+	if engine.cfg.TaskCPUMemLimit {
+		if _, ok := supportedVersions[dockerclient.Version_1_22]; ok {
+			capabilities = append(capabilities, capabilityPrefix+capabilityTaskCPUMemLimit)
```
Comment (aaithal) [inline agent/engine/docker_task_engine.go]: this shouldn't use `capabilityPrefix`. It should use `"ecs.capability."` instead (refer https://github.com/aws/amazon-ecs-agent/blob/enis/agent/app/agent_capability.go#L29)

File: agent/engine/docker_task_engine.go [hunk 13]
```diff
@@ -830,6 +832,16 @@ func (engine *DockerTaskEngine) Capabilities() []string {
...
+		} else {
+			seelog.Warn("Task CPU + Mem Limit disabled due to unsupported Docker version")
+			engine.cfg.TaskCPUMemLimit = false
```
Comment (aaithal) [inline agent/engine/docker_task_engine.go]: The approach we've taken recently is to kill the agent with a terminal error code if it can't register with the required capability when asked to do so via the config. We should rethink if we have to take the same approach here. The argument is that it's better to not register the instance than register it without the capability that was asked for as it leads to a more consistent troubleshooting experience.

Comment (petderek) [reply agent/engine/docker_task_engine.go]: I think its slightly different here. In this case, 'enabled' is the default and it can be manually overridden to false.

Since it is a default, it would be disruptive to not register the instance (in case of updates, old versions of docker, etc).

Comment (petderek) [reply agent/engine/docker_task_engine.go]: After looking over some things further we think that the best way to handle this is to make enhancements to the config package. We need to fail if the value was set explicitly, but we should only fall back to old behavior gracefully if the value was derived from a default. 

Its a significant change, so I'd like to do that on a separate PR and keep progress moving forward on this branch.

File: agent/engine/docker_task_engine.go [hunk 14]
```diff
@@ -45,7 +45,9 @@ const (
 	capabilityPrefix             = "com.amazonaws.ecs.capability."
 	capabilityTaskIAMRole        = "task-iam-role"
 	capabilityTaskIAMRoleNetHost = "task-iam-role-network-host"
+	capabilityTaskCPUMemLimit    = "task-cpu-mem-limit"
 	labelPrefix                  = "com.amazonaws.ecs."
+	attributePrefix              = "ecs.capability."
```
Comment (aaithal) [inline agent/engine/docker_task_engine.go]: If/when you rebase from `dev`, this is going to cause a merge conflict. I'd suggest that soon after merging this PR as the capabilities code has been moved to "agent/app/agent_capabilities.go"

File: agent/config/config.go [hunk 15]
```diff
@@ -288,6 +288,7 @@ func environmentConfig() (Config, error) {
 	appArmorCapable := utils.ParseBool(os.Getenv("ECS_APPARMOR_CAPABLE"), false)
 	taskIAMRoleEnabled := utils.ParseBool(os.Getenv("ECS_ENABLE_TASK_IAM_ROLE"), false)
 	taskIAMRoleEnabledForNetworkHost := utils.ParseBool(os.Getenv("ECS_ENABLE_TASK_IAM_ROLE_NETWORK_HOST"), false)
+	taskCPUMemLimitEnabled := utils.ParseBool(os.Getenv("ECS_ENABLE_TASK_CPU_MEM_LIMIT"), true)
```
Comment (richardpen) [inline agent/config/config.go]: Can you add this to the readme file?
